README.md
# Rust AD983x Low-Power Programmable Waveform Generator / Direct Digital Synthesizer (DDS) Driver

[![crates.io](https://img.shields.io/crates/v/ad983x.svg)](https://crates.io/crates/ad983x)
[![Docs](https://docs.rs/ad983x/badge.svg)](https://docs.rs/ad983x)
![Minimum Supported Rust Version](https://img.shields.io/badge/rustc-1.54+-blue.svg)
[![Build Status](https://github.com/eldruin/ad983x-rs/workflows/Build/badge.svg)](https://github.com/eldruin/ad983x-rs/actions?query=workflow%3ABuild)
[![Coverage Status](https://coveralls.io/repos/github/eldruin/ad983x-rs/badge.svg?branch=master)](https://coveralls.io/github/eldruin/ad983x-rs?branch=master)

This is a platform agnostic Rust driver for the AD9833, AD9834, AD9837 and AD9838 low-power programmable waveform generators / direct digital synthesizers (DDS) using the [`embedded-hal`] traits.

This driver allows you to:
- Enable/disable/reset the device. See `enable()`.
- Set the frequency registers. See: `set_frequency()`.
- Select the output frequency register. See: `select_frequency()`.
- Set the phase registers. See: `set_phase()`.
- Select the output phase register. See: `select_phase()`.
- Set the frequency registers MSBs/LSBs separately. See: `set_frequency_msb()`.
- Set the output waveform. See: `set_output_waveform()`.
- Power down/up device parts. See: `set_powered_down()`.
- Select control source on AD9834/AD9838. See: `set_control_source()`.

[Introductory blog post](https://blog.eldruin.com/ad983x-waveform-generator-dds-driver-in-rust/)

## The devices

The AD9833, AD9834, AD9837 and AD9838 are low power, programmable waveform generators capable of producing sine, triangular, and square wave outputs. Waveform generation is required in various types of sensing, actuation, and time domain reflectometry (TDR) applications. The output frequency and phase are software programmable, allowing easy tuning. No external components are needed. The frequency registers are 28 bits wide: with a 25 MHz clock rate, resolution of 0.1 Hz can be achieved; with a 1 MHz clock rate, the AD9833 can be tuned to 0.004 Hz resolution.

The devices are written to via a 3-wire serial interface (SPI). This serial interface operates at clock rates up to 40 MHz and is compatible with DSP and microcontroller standards. The devices operate with a power supply from 2.3 V to 5.5 V.

Datasheets:
- [AD9833](https://www.analog.com/media/en/technical-documentation/data-sheets/ad9833.PDF)
- [AD9834](https://www.analog.com/media/en/technical-documentation/data-sheets/AD9834.PDF)
- [AD9837](https://www.analog.com/media/en/technical-documentation/data-sheets/AD9837.PDF)
- [AD9838](https://www.analog.com/media/en/technical-documentation/data-sheets/AD9838.PDF)

Application Note:
- [Programming the AD9833/AD9834](https://www.analog.com/media/en/technical-documentation/application-notes/AN-1070.pdf)

Article explaining DDS using an AD9833:
- [All about direct digital synthesis](https://www.analog.com/en/analog-dialogue/articles/all-about-direct-digital-synthesis.html)

## Usage

To use this driver, import this crate and an `embedded_hal` implementation,
then instantiate the appropriate device.

I wrote an example MIDI player that plays Beethoven's ninth symphony in hardware :). See: [driver-examples].

[driver-examples]: https://github.com/eldruin/driver-examples

```rust
use ad983x::{Ad983x, FrequencyRegister};
use embedded_hal::spi::blocking::ExclusiveDevice;
use linux_embedded_hal::{Spidev, SysfsPin};

fn main() {
    let spi = Spidev::open("/dev/spidev0.0").unwrap();
    let chip_select = SysfsPin::new(25);
    let dev = ExclusiveDevice::new(spi, chip_select);
    let mut dds = Ad983x::new_ad9833(dev);
    dds.reset().unwrap(); // reset is necessary before operation
    dds.set_frequency(FrequencyRegister::F0, 4724).unwrap();
    dds.enable().unwrap();
    // Given a 25 MHz clock, this now outputs a sine wave
    // with a frequency of 440 Hz, which is a standard
    // A4 tone.

    // Get device back
    let _dev = dds.destroy();
}
```

## Status

- [X] Compatible with AD9833/AD9837
- [X] Compatible with AD9834/AD9838
- [ ] Compatible with AD9832/AD9835

## Support

For questions, issues, feature requests like compatibility with similar devices
and other changes, please file an
[issue in the github project](https://github.com/eldruin/ad983x-rs/issues).

## License

Licensed under either of

 * Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or
   http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or
   http://opensource.org/licenses/MIT)

at your option.

### Contributing

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall
be dual licensed as above, without any additional terms or conditions.

[`embedded-hal`]: https://github.com/rust-embedded/embedded-hal


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Changed
- [breaking-change] Updated `embedded-hal` to version `1.0.0-alpha.8`.
- Updated MSRV to Rust 1.54.0.
- Updated dependencies.

## [0.3.0] - 2021-09-24

### Changed
- [breaking-change] Remove `Default` implementation for `Ad989x`.
- Updated dependencies.

## [0.2.0] - 2019-11-10

### Changed
- Use `embedded_hal::digital::v2::OutputPin` traits for SPI chip-select pin.

## [0.1.1] - 2019-05-30

### Added
- Support for AD9834 and AD9838 devices.

## 0.1.0 - 2019-05-29

Initial release to crates.io. AD9833 and AD9837 devices are fully supported.
All changes will be documented in this CHANGELOG.

[Unreleased]: https://github.com/eldruin/ad983x-rs/compare/v0.3.0...HEAD
[0.3.0]: https://github.com/eldruin/ad983x-rs/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/eldruin/ad983x-rs/compare/v0.1.1...v0.2.0
[0.1.1]: https://github.com/eldruin/ad983x-rs/compare/v0.1.0...v0.1.1

tests/base/mod.rs
use ad983x::{marker, Ad983x};
use embedded_hal_mock::spi::{Mock as SpiMock, Transaction as SpiTrans};

pub struct BitFlags;
impl BitFlags {
    pub const D15: u8 = 1 << 7;
    pub const D14: u8 = 1 << 6;
    pub const D13: u8 = 1 << 5;
    pub const B28: u8 = 1 << 5;
    pub const HLB: u8 = 1 << 4;
    pub const FSELECT: u8 = 1 << 3;
    pub const PSELECT: u8 = 1 << 2;
    pub const PIN_SW: u8 = 1 << 1;
    pub const SLEEP_MCLK: u8 = 1 << 7;
    pub const SLEEP_DAC: u8 = 1 << 6;
    pub const OPBITEN: u8 = 1 << 5;
    pub const SIGN_PIB: u8 = 1 << 4;
    pub const DIV2: u8 = 1 << 3;
    pub const MODE: u8 = 1 << 1;
    pub const RESET: u8 = 1;
    pub const FREQ0: u8 = 1 << 6;
    pub const FREQ1: u8 = 1 << 7;
}

pub fn new_ad9833(transactions: &[SpiTrans]) -> Ad983x<SpiMock, marker::Ad9833Ad9837> {
    let wrapped: Vec<SpiTrans> = transactions
        .iter()
        .flat_map(|trans| {
            [
                SpiTrans::transaction_start(),
                trans.clone(),
                SpiTrans::transaction_end(),
            ]
        })
        .collect();
    Ad983x::new_ad9833(SpiMock::new(wrapped.iter()))
}

pub fn new_ad9834(transactions: &[SpiTrans]) -> Ad983x<SpiMock, marker::Ad9834Ad9838> {
    let wrapped: Vec<SpiTrans> = transactions
        .iter()
        .flat_map(|trans| {
            [
                SpiTrans::transaction_start(),
                trans.clone(),
                SpiTrans::transaction_end(),
            ]
        })
        .collect();
    Ad983x::new_ad9834(SpiMock::new(wrapped.iter()))
}

pub fn new_ad9837(transactions: &[SpiTrans]) -> Ad983x<SpiMock, marker::Ad9833Ad9837> {
    let wrapped: Vec<SpiTrans> = transactions
        .iter()
        .flat_map(|trans| {
            [
                SpiTrans::transaction_start(),
                trans.clone(),
                SpiTrans::transaction_end(),
            ]
        })
        .collect();
    Ad983x::new_ad9837(SpiMock::new(wrapped.iter()))
}

pub fn new_ad9838(transactions: &[SpiTrans]) -> Ad983x<SpiMock, marker::Ad9834Ad9838> {
    let wrapped: Vec<SpiTrans> = transactions
        .iter()
        .flat_map(|trans| {
            [
                SpiTrans::transaction_start(),
                trans.clone(),
                SpiTrans::transaction_end(),
            ]
        })
        .collect();
    Ad983x::new_ad9838(SpiMock::new(wrapped.iter()))
}

pub fn destroy<IC>(device: Ad983x<SpiMock, IC>) {
    device.destroy().done();
}


tests/integration.rs
use ad983x::{
    ControlSource, FrequencyRegister as FreqReg, OutputWaveform as OW, PhaseRegister as PhaseReg,
    PoweredDown as PD, SignBitOutput as SBO,
};
use embedded_hal_mock::spi::Transaction as SpiTrans;

mod base;
use crate::base::{destroy, new_ad9833, new_ad9834, new_ad9837, new_ad9838, BitFlags as BF};

#[test]
fn can_create_and_destroy_ad9833() {
    let dev = new_ad9833(&[]);
    destroy(dev);
}

#[test]
fn can_create_and_destroy_ad9837() {
    let dev = new_ad9837(&[]);
    destroy(dev);
}

#[test]
fn can_create_and_destroy_ad9834() {
    let dev = new_ad9834(&[]);
    destroy(dev);
}

#[test]
fn can_create_and_destroy_ad9838() {
    let dev = new_ad9838(&[]);
    destroy(dev);
}

#[test]
fn can_enable() {
    let transitions = [SpiTrans::write_vec(vec![0, 0])];
    let mut dev = new_ad9833(&transitions);
    dev.enable().unwrap();
    destroy(dev);
}

#[test]
fn can_disable() {
    let transitions = [SpiTrans::write_vec(vec![BF::RESET, 0])];
    let mut dev = new_ad9833(&transitions);
    dev.disable().unwrap();
    destroy(dev);
}

#[test]
fn can_reset() {
    let transitions = [SpiTrans::write_vec(vec![BF::RESET, 0])];
    let mut dev = new_ad9833(&transitions);
    dev.reset().unwrap();
    destroy(dev);
}

#[test]
fn cannot_set_too_fast_frequency() {
    let mut dev = new_ad9833(&[]);
    dev.set_frequency(FreqReg::F0, 1 << 28)
        .expect_err("Should return error");
    destroy(dev);
}

#[test]
fn can_set_freq0() {
    let transitions = [
        SpiTrans::write_vec(vec![BF::B28 | BF::RESET, 0]),
        SpiTrans::write_vec(vec![BF::FREQ0 | 0xD, 0xEF]),
        SpiTrans::write_vec(vec![BF::FREQ0 | 0x26, 0xAF]),
    ];
    let mut dev = new_ad9833(&transitions);
    dev.set_frequency(FreqReg::F0, 0x9AB_CDEF).unwrap();
    destroy(dev);
}

#[test]
fn can_set_freq1() {
    let transitions = [
        SpiTrans::write_vec(vec![BF::B28 | BF::RESET, 0]),
        SpiTrans::write_vec(vec![BF::FREQ1 | 0xD, 0xEF]),
        SpiTrans::write_vec(vec![BF::FREQ1 | 0x26, 0xAF]),
    ];
    let mut dev = new_ad9833(&transitions);
    dev.set_frequency(FreqReg::F1, 0x9AB_CDEF).unwrap();
    destroy(dev);
}

#[test]
fn can_select_freq0() {
    let transitions = [SpiTrans::write_vec(vec![BF::RESET, 0])];
    let mut dev = new_ad9833(&transitions);
    dev.select_frequency(FreqReg::F0).unwrap();
    destroy(dev);
}

#[test]
fn can_select_freq1() {
    let transitions = [SpiTrans::write_vec(vec![BF::FSELECT | BF::RESET, 0])];
    let mut dev = new_ad9833(&transitions);
    dev.select_frequency(FreqReg::F1).unwrap();
    destroy(dev);
}

#[test]
fn cannot_set_wrong_phase() {
    let mut dev = new_ad9833(&[]);
    dev.set_phase(PhaseReg::P0, 1 << 12)
        .expect_err("Should return error");
    destroy(dev);
}

#[test]
fn can_set_phase0() {
    let transitions = [SpiTrans::write_vec(vec![BF::D15 | BF::D14 | 0xA, 0xBC])];
    let mut dev = new_ad9833(&transitions);
    dev.set_phase(PhaseReg::P0, 0xABC).unwrap();
    destroy(dev);
}

#[test]
fn can_set_phase1() {
    let transitions = [SpiTrans::write_vec(vec![
        BF::D15 | BF::D14 | BF::D13 | 0xA,
        0xBC,
    ])];
    let mut dev = new_ad9833(&transitions);
    dev.set_phase(PhaseReg::P1, 0xABC).unwrap();
    destroy(dev);
}

#[test]
fn can_select_phase0() {
    let transitions = [SpiTrans::write_vec(vec![BF::RESET, 0])];
    let mut dev = new_ad9833(&transitions);
    dev.select_phase(PhaseReg::P0).unwrap();
    destroy(dev);
}

#[test]
fn can_select_phase1() {
    let transitions = [SpiTrans::write_vec(vec![BF::PSELECT | BF::RESET, 0])];
    let mut dev = new_ad9833(&transitions);
    dev.select_phase(PhaseReg::P1).unwrap();
    destroy(dev);
}

macro_rules! ow_test {
    ($name:ident, $create:ident, $ow:ident, $control:expr) => {
        #[test]
        fn $name() {
            let transitions = [SpiTrans::write_vec(vec![BF::RESET, $control])];
            let mut dev = $create(&transitions);
            dev.set_output_waveform(OW::$ow).unwrap();
            destroy(dev);
        }
    };
}

ow_test!(can_set_sinusoidal_out, new_ad9833, Sinusoidal, 0);
ow_test!(can_set_triangle_out, new_ad9833, Triangle, BF::MODE);
ow_test!(
    can_set_sq_msb_out,
    new_ad9833,
    SquareMsbOfDac,
    BF::OPBITEN | BF::DIV2
);
ow_test!(
    can_set_sq_msb_div2_out,
    new_ad9833,
    SquareMsbOfDacDiv2,
    BF::OPBITEN
);
ow_test!(can_set_sinusoidal_out_ad9838, new_ad9838, Sinusoidal, 0);
ow_test!(can_set_triangle_out_ad9838, new_ad9838, Triangle, BF::MODE);

macro_rules! pd_test {
    ($name:ident, $pd:ident, $control:expr) => {
        #[test]
        fn $name() {
            let transitions = [SpiTrans::write_vec(vec![BF::RESET, $control])];
            let mut dev = new_ad9833(&transitions);
            dev.set_powered_down(PD::$pd).unwrap();
            destroy(dev);
        }
    };
}

pd_test!(can_set_pd_nothing, Nothing, 0);
pd_test!(can_set_pd_dac, Dac, BF::SLEEP_DAC);
pd_test!(can_set_pd_mclk, InternalClock, BF::SLEEP_MCLK);
pd_test!(
    can_set_pd_dac_mclk,
    DacAndInternalClock,
    BF::SLEEP_MCLK | BF::SLEEP_DAC
);

#[test]
fn cannot_set_wrong_freq_msb() {
    let mut dev = new_ad9833(&[]);
    dev.set_frequency_msb(FreqReg::F0, 1 << 14)
        .expect_err("Should return error");
    destroy(dev);
}

#[test]
fn cannot_set_wrong_freq_lsb() {
    let mut dev = new_ad9833(&[]);
    dev.set_frequency_lsb(FreqReg::F0, 1 << 14)
        .expect_err("Should return error");
    destroy(dev);
}

#[test]
fn can_set_freq_msb() {
    let transitions = [
        SpiTrans::write_vec(vec![BF::HLB | BF::RESET, 0]),
        SpiTrans::write_vec(vec![BF::FREQ0 | 0xD, 0xEF]),
    ];
    let mut dev = new_ad9833(&transitions);
    dev.set_frequency_msb(FreqReg::F0, 0xDEF).unwrap();
    destroy(dev);
}

#[test]
fn can_set_freq_lsb() {
    let transitions = [SpiTrans::write_vec(vec![BF::FREQ1 | 0xD, 0xEF])];
    let mut dev = new_ad9833(&transitions);
    dev.set_frequency_lsb(FreqReg::F1, 0xDEF).unwrap();
    destroy(dev);
}

#[test]
fn cannot_set_ow_dac_ad9838() {
    let mut dev = new_ad9838(&[]);
    dev.set_output_waveform(OW::SquareMsbOfDac)
        .expect_err("Should return error");
    destroy(dev);
}

#[test]
fn cannot_set_ow_dac_div2_ad9838() {
    let mut dev = new_ad9838(&[]);
    dev.set_output_waveform(OW::SquareMsbOfDacDiv2)
        .expect_err("Should return error");
    destroy(dev);
}

macro_rules! sbo_test {
    ($name:ident, $sbo:ident, $control:expr) => {
        #[test]
        fn $name() {
            let transitions = [SpiTrans::write_vec(vec![BF::RESET, $control])];
            let mut dev = new_ad9838(&transitions);
            dev.set_sign_bit_output(SBO::$sbo).unwrap();
            destroy(dev);
        }
    };
}

sbo_test!(can_set_disabled_sign_out, Disabled, 0);
sbo_test!(
    can_set_comp_sign_out,
    Comparator,
    BF::OPBITEN | BF::SIGN_PIB | BF::DIV2
);
sbo_test!(
    can_set_sq_msb_sign_out,
    SquareMsbOfDac,
    BF::OPBITEN | BF::DIV2
);
sbo_test!(
    can_set_sq_msb_div2_sign_out,
    SquareMsbOfDacDiv2,
    BF::OPBITEN
);

#[test]
fn can_set_control_source_sw() {
    let transitions = [SpiTrans::write_vec(vec![BF::RESET, 0])];
    let mut dev = new_ad9838(&transitions);
    dev.set_control_source(ControlSource::Software).unwrap();
    destroy(dev);
}

#[test]
fn can_set_control_source_hw_pins() {
    let transitions = [SpiTrans::write_vec(vec![BF::RESET | BF::PIN_SW, 0])];
    let mut dev = new_ad9838(&transitions);
    dev.set_control_source(ControlSource::HardwarePins).unwrap();
    destroy(dev);
}


src/ad9833_ad9837.rs
use embedded_hal::spi::blocking::{SpiBus, SpiDevice};

use crate::{marker, Ad983x, BitFlags, Error, OutputWaveform};

impl<DEV, E> Ad983x<DEV, marker::Ad9833Ad9837>
where
    DEV: SpiDevice<Error = E>,
    DEV::Bus: SpiBus,
{
    /// Create a new instance of an AD9833 device.
    ///
    /// Remember to call `reset()` before using the device after power up.
    pub fn new_ad9833(spi: DEV) -> Self {
        Self::create(spi)
    }
    /// Create a new instance of an AD9837 device.
    ///
    /// Remember to call `reset()` before using the device after power up.
    pub fn new_ad9837(spi: DEV) -> Self {
        // Behaves the same as AD9833
        Self::create(spi)
    }

    /// Set the output waveform
    pub fn set_output_waveform(&mut self, waveform: OutputWaveform) -> Result<(), Error<E>> {
        let control = match waveform {
            OutputWaveform::Sinusoidal => self
                .control
                .with_low(BitFlags::OPBITEN)
                .with_low(BitFlags::MODE),
            OutputWaveform::Triangle => self
                .control
                .with_low(BitFlags::OPBITEN)
                .with_high(BitFlags::MODE),
            OutputWaveform::SquareMsbOfDac => self
                .control
                .with_high(BitFlags::OPBITEN)
                .with_low(BitFlags::MODE)
                .with_high(BitFlags::DIV2),
            OutputWaveform::SquareMsbOfDacDiv2 => self
                .control
                .with_high(BitFlags::OPBITEN)
                .with_low(BitFlags::MODE)
                .with_low(BitFlags::DIV2),
        };
        self.write_control(control)
    }
}


src/lib.rs
//! This is a platform agnostic Rust driver for the AD9833, AD9834, AD9837
//! and AD9838 low-power programmable waveform generators / direct digital
//! synthesizers (DDS) using the [`embedded-hal`] traits.
//!
//! [`embedded-hal`]: https://github.com/rust-embedded/embedded-hal
//!
//! This driver allows you to:
//! - Enable/disable/reset the device. See [`enable()`].
//! - Set the frequency registers. See: [`set_frequency()`].
//! - Select the output frequency register. See: [`select_frequency()`].
//! - Set the phase registers. See: [`set_phase()`].
//! - Select the output phase register. See: [`select_phase()`].
//! - Set the frequency registers MSBs/LSBs separately. See: [`set_frequency_msb()`].
//! - Set the output waveform. See: [`set_output_waveform()`].
//! - Power down/up device parts. See: [`set_powered_down()`].
//! - Select control source on AD9834/AD9838. See: [`set_control_source()`].
//!
//! [`enable()`]: struct.Ad983x.html#method.enable
//! [`set_frequency()`]: struct.Ad983x.html#method.set_frequency
//! [`select_frequency()`]: struct.Ad983x.html#method.select_frequency
//! [`set_phase()`]: struct.Ad983x.html#method.set_phase
//! [`select_phase()`]: struct.Ad983x.html#method.select_phase
//! [`set_frequency_msb()`]: struct.Ad983x.html#method.set_frequency_msb
//! [`set_output_waveform()`]: struct.Ad983x.html#method.set_output_waveform
//! [`set_powered_down()`]: struct.Ad983x.html#method.set_powered_down
//! [`set_control_source()`]: struct.Ad983x.html#method.set_control_source
//!
//! [Introductory blog post](https://blog.eldruin.com/ad983x-waveform-generator-dds-driver-in-rust/)
//!
//! ## The devices
//!
//! The AD9833, AD9834, AD9837 and AD9838 are low power, programmable waveform
//! generators capable of producing sine, triangular, and square wave outputs.
//! Waveform generation is required in various types of sensing, actuation,
//! and time domain reflectometry (TDR) applications. The output frequency and
//! phase are software programmable, allowing easy tuning. No external
//! components are needed. The frequency registers are 28 bits wide: with a
//! 25 MHz clock rate, resolution of 0.1 Hz can be achieved; with a 1 MHz
//! clock rate, the AD9833 can be tuned to 0.004 Hz resolution.
//!
//! The devices are written to via a 3-wire serial interface (SPI).
//! This serial interface operates at clock rates up to 40 MHz and is
//! compatible with DSP and microcontroller standards. The devices operate
//! with a power supply from 2.3 V to 5.5 V.
//!
//! Datasheets:
//! - [AD9833](https://www.analog.com/media/en/technical-documentation/data-sheets/ad9833.PDF)
//! - [AD9834](https://www.analog.com/media/en/technical-documentation/data-sheets/AD9834.PDF)
//! - [AD9837](https://www.analog.com/media/en/technical-documentation/data-sheets/AD9837.PDF)
//! - [AD9838](https://www.analog.com/media/en/technical-documentation/data-sheets/AD9838.PDF)
//!
//! Application Note:
//! - [Programming the AD9833/AD9834](https://www.analog.com/media/en/technical-documentation/application-notes/AN-1070.pdf)
//!
//! Article explaining DDS using an AD9833:
//! - [All about direct digital synthesis](https://www.analog.com/en/analog-dialogue/articles/all-about-direct-digital-synthesis.html)
//!
//! ## SysfsPin / Software control source on AD9834/AD9838
//!
//! AD9834/AD9838 devices offer the possibility to control several functions
//! either through hardware pins or software settings. While hardware pin
//! control is selected, these software operations will be ignored by the hardware.
//! This driver allows this as well as it would be a valid use case to
//! configure the status of these functions while on hardware pin control mode
//! in preparation for a smooth switch to software control.
//!
//! ## Usage examples (see also examples folder)
//!
//! To use this driver, import this crate and an `embedded_hal` implementation,
//! then instantiate the appropriate device.
//! In the following examples an instance of the device AD9833 will be created
//! as an example. Other devices can be created with similar methods like:
//! `Ad983x::new_ad9837(...)`.
//!
//! Please find additional examples using hardware in this repository: [driver-examples].
//!
//! This includes an example MIDI player that plays Beethoven's ninth symphony.
//!
//! [driver-examples]: https://github.com/eldruin/driver-examples
//!
//! ### Set the frequency register 0 and enable
//!
//! ```no_run
//! use ad983x::{Ad983x, FrequencyRegister};
//! use embedded_hal::spi::blocking::ExclusiveDevice;
//! use linux_embedded_hal::{SysfsPin, Spidev};
//!
//! let spi = Spidev::open("/dev/spidev0.0").unwrap();
//! let chip_select = SysfsPin::new(25);
//! let dev = ExclusiveDevice::new(spi, chip_select);
//! let mut dds = Ad983x::new_ad9833(dev);
//! dds.reset().unwrap(); // reset is necessary before operation
//! dds.set_frequency(FrequencyRegister::F0, 4724).unwrap();
//! dds.enable().unwrap();
//! // Given a 25 MHz clock, this now outputs a sine wave
//! // with a frequency of 440 Hz, which is a standard
//! // A4 tone.
//!
//! // Get device back
//! let _dev = dds.destroy();
//! ```
//!
//! ### Set frequency registers 0 and 1 and alternate between them
//!
//! With a 25 MHz clock this alternates between A4 and D5 tones.
//!
//! ```no_run
//! use ad983x::{Ad983x, FrequencyRegister};
//! use embedded_hal::spi::blocking::ExclusiveDevice;
//! use linux_embedded_hal::{SysfsPin, Spidev};
//!
//! let spi = Spidev::open("/dev/spidev0.0").unwrap();
//! let chip_select = SysfsPin::new(25);
//! let dev = ExclusiveDevice::new(spi, chip_select);
//! let mut dds = Ad983x::new_ad9833(dev);
//! dds.reset().unwrap(); // reset is necessary before operation
//! // A4 tone for a 25 MHz clock
//! dds.set_frequency(FrequencyRegister::F0, 4724).unwrap();
//! // D5 tone for a 25 MHz clock
//! dds.set_frequency(FrequencyRegister::F1, 6306).unwrap();
//! dds.enable().unwrap();
//! loop {
//!     // some delay
//!     dds.select_frequency(FrequencyRegister::F1).unwrap();
//!     // some delay
//!     dds.select_frequency(FrequencyRegister::F0).unwrap();
//! }
//! ```
//!
//! ### Set the phase register 1 and select it
//!
//! ```no_run
//! use ad983x::{Ad983x, PhaseRegister};
//! use embedded_hal::spi::blocking::ExclusiveDevice;
//! use linux_embedded_hal::{SysfsPin, Spidev};
//!
//! let spi = Spidev::open("/dev/spidev0.0").unwrap();
//! let chip_select = SysfsPin::new(25);
//! let dev = ExclusiveDevice::new(spi, chip_select);
//! let mut dds = Ad983x::new_ad9833(dev);
//! dds.reset().unwrap(); // reset is necessary before operation
//! dds.set_phase(PhaseRegister::P1, 4724).unwrap();
//! dds.select_phase(PhaseRegister::P1).unwrap();
//! ```
//!
//! ### Set output waveform to be triangular
//!
//! ```no_run
//! use ad983x::{Ad983x, OutputWaveform};
//! use embedded_hal::spi::blocking::ExclusiveDevice;
//! use linux_embedded_hal::{SysfsPin, Spidev};
//!
//! let spi = Spidev::open("/dev/spidev0.0").unwrap();
//! let chip_select = SysfsPin::new(25);
//! let dev = ExclusiveDevice::new(spi, chip_select);
//! let mut dds = Ad983x::new_ad9833(dev);
//! dds.reset().unwrap(); // reset is necessary before operation
//! dds.set_output_waveform(OutputWaveform::Triangle).unwrap();
//! ```
//!
//! ### Power down the DAC
//!
//! ```no_run
//! use ad983x::{Ad983x, PoweredDown};
//! use embedded_hal::spi::blocking::ExclusiveDevice;
//! use linux_embedded_hal::{SysfsPin, Spidev};
//!
//! let spi = Spidev::open("/dev/spidev0.0").unwrap();
//! let chip_select = SysfsPin::new(25);
//! let dev = ExclusiveDevice::new(spi, chip_select);
//! let mut dds = Ad983x::new_ad9833(dev);
//! dds.reset().unwrap(); // reset is necessary before operation
//! dds.set_powered_down(PoweredDown::Dac).unwrap();
//! ```
//!
//! ### Use hardware pins as control source
//!
//! ```no_run
//! use ad983x::{Ad983x, ControlSource};
//! use embedded_hal::spi::blocking::ExclusiveDevice;
//! use linux_embedded_hal::{SysfsPin, Spidev};
//!
//! let spi = Spidev::open("/dev/spidev0.0").unwrap();
//! let chip_select = SysfsPin::new(25);
//! let dev = ExclusiveDevice::new(spi, chip_select);
//! let mut dds = Ad983x::new_ad9838(dev);
//! dds.reset().unwrap(); // reset is necessary before operation
//! dds.set_control_source(ControlSource::HardwarePins).unwrap();
//! // Hardware pins can now be used to control the device.
//! // The corresponding software settings will be ignored.
//! ```

#![deny(unsafe_code, missing_docs)]
#![no_std]

use core::marker::PhantomData;
use embedded_hal::spi::{Mode, MODE_2};

/// All possible errors in this crate
#[derive(Debug)]
pub enum Error<E> {
    /// SPI communication error
    Spi(E),
    /// Invalid argument provided
    InvalidArgument,
}

/// Frequency registers
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum FrequencyRegister {
    /// Frequency register 0
    F0,
    /// Frequency register 1
    F1,
}

/// Phase registers
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum PhaseRegister {
    /// Phase register 0
    P0,
    /// Phase register 1
    P1,
}

/// Output waveform
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum OutputWaveform {
    /// Sinusoidal wave (default)
    Sinusoidal,
    /// Triangle wave
    Triangle,
    /// Square wave with its value matching the MSB of DAC data
    /// (not available on AD9834/AD9838, use `SignBitOutput`)
    SquareMsbOfDac,
    /// Square wave with its value matching the MSB of DAC data divided by 2
    /// (not available on AD9834/AD9838, use `SignBitOutput`)
    SquareMsbOfDacDiv2,
}

/// Sign bit output on AD9834/AD9838 devices
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum SignBitOutput {
    /// Disabled (high impedance) (default)
    Disabled,
    /// Comparator output
    Comparator,
    /// Square wave with its value matching the MSB of DAC data
    SquareMsbOfDac,
    /// Square wave with its value matching the MSB of DAC data divided by 2
    SquareMsbOfDacDiv2,
}

/// Powered-down device configuration
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum PoweredDown {
    /// All chip parts are enabled (default)
    Nothing,
    /// Power down only the DAC
    Dac,
    /// Disable only the internal clock
    InternalClock,
    /// Power down the DAC and disable the internal clock
    DacAndInternalClock,
}

/// Hardware pin / software control source for the functions:
/// frequency register selection, phase register selection,
/// reset of internal registers, and DAC power-down.
/// (Only available on AD9834 and AD9838 devices)
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum ControlSource {
    /// Functions are controlled only through software (default)
    Software,
    /// Functions are controlled only through hardware pins
    HardwarePins,
}

/// SPI mode (CPOL = 1, CPHA = 0)
pub const MODE: Mode = MODE_2;

/// Markers
#[doc(hidden)]
pub mod marker {
    /// AD9833/AD9837 device
    pub enum Ad9833Ad9837 {}
    /// AD9834/AD9838 device
    pub enum Ad9834Ad9838 {}
}

struct BitFlags;

#[derive(Debug, Default, Clone, Copy, PartialEq)]
struct Config {
    bits: u16,
}

/// AD983x direct digital synthesizer
#[derive(Debug)]
pub struct Ad983x<DEV, IC> {
    spi: DEV,
    control: Config,
    _ic: PhantomData<IC>,
}

mod ad9833_ad9837;
mod ad9834_ad9838;
mod common;

mod private {
    use super::marker;
    pub trait Sealed {}

    impl Sealed for marker::Ad9833Ad9837 {}
    impl Sealed for marker::Ad9834Ad9838 {}
}
